### ISSUE TITLE
Navigation Blocking doesn't work with browser's back button.

### ISSUE THREAD

**Issue by baki-kucukcakiroglu** [TanStack/router#701]
### Describe the bug

Navigation blocking works with Link or useNavigate but doesn't work with browser's back button. So if someone go back using back button, blocking is being surpassed. 

### Your Example Website or App

https://codesandbox.io/s/boring-buck-nfxmgj

### Steps to Reproduce the Bug or Issue

Note: Actually I'm using "@tanstack/react-router": "^0.0.1-beta.171" but this version was not an option in sandbox. 

Example application has two pages called FirstPage and SecondPage. There is a Link component on each page that allows navigate between pages. The Link on the FirstPage routes to SecondPage and vice versa. Navigating to FirstPage from SecondPage has been blocked by useBlocker. If Link is clicked, blocker works well but if back button of browser is used blocker fails. 

### Expected behavior

Normally, useBlocker should also block routings by back button of a browser.

### Platform

- OS: [macOS]
- Browser: [Chrome]

**Comment by lsmurray** [comment]
Getting the same bug, also getting infinite loops with the blocker. If I press `ok` on the initial click I can leave the page but if I press `cancel` then the blocker shows up again and again and again. Let me know if you need a minimum reproducible example.

**Comment by tannerlinsley** [comment]
Looking into this.

**Comment by baki-kucukcakiroglu** [comment]
are there any improvements?

**Comment by jgarciamccausland** [comment]
I believe this is fixed and can be closed.

**Comment by baki-kucukcakiroglu** [comment]
@jgarciamccausland In my opinion, it is not solved yet.

**Comment by Jarzka** [comment]
This bug still happens. Any news?

**Comment by tannerlinsley** [comment]
As of the latest version, we're controlling as much of the browser as possible. The back button AFAIK is not cancellable via current platform/browser APIs via a popState event with something like `onbeforeunload` (since the page isn't actually unloading). 

We do monkey patch the history.popState API, but clicking the actual browser back button is wired directly into the platform and is not something we can get between. It may be possible with the new Navigation API, but is something we would need to explore more (https://developer.chrome.com/docs/web-platform/navigation-api).

Closing, since there is nothing more we can technically do.

**Comment by Jarzka** [comment]
It seems that the popstate event cannot be prevent, but I have seen some people using this kind of method to prevent the navigation:

```ts
    window.onpopstate = function () {
        history.go(1);
    };
```

See: https://codepen.io/dhavalt10/pen/rGLBzB

I quickly tried this method in our application using TanStack router and it seems to prevent the navigation (actually the url changes and then switches back, but the point is that I do not lose any data)

**Comment by tannerlinsley** [comment]
This feels extremely hacky. Does it work consistently across major browsers?

**Comment by Jarzka** [comment]
Actually, it doesn't. Based on my tests, the blocking method works once in Chrome and on Firefox it does not work at all. So I guess it's not a good solution. Plus, like you said, it feels very hacky.

**Comment by tannerlinsley** [comment]
Sounds good. Maybe that's something we could add to an FAQ or something and let people decide if they want to go down that path...

**Comment by ollejernstrom** [comment]
@tannerlinsley 

I was just wondering how it could be a difference between how `react-router` does it since there it works as expected by default? 

Please see https://codesandbox.io/p/sandbox/react-router-6-use-prompt-use-blocker-17r2f?file=%2Findex.js

**Comment by rothsandro** [comment]
React Router supports blocking the navigation by reverting the navigation[^1][^2]:

> In v6, non-blocked navigations are a no-op, and blocked navigations are immediately reverted which re-syncs the with the URL before any other navigations can happen.

I agree it's a workaround but normal navigation can already be blocked and for the user there is not really a difference between page reload / link navigation / back navigation, so it would be nice to have the same behavior for all navigations.

Are you willing to implement a similar solution? This blocks us and likely other teams from using TanStack Router as we have to requirement to handle dirty state for all navigations.

[^1]: https://github.com/remix-run/react-router/blob/main/decisions/0001-use-blocker.md#how-can-we-tackle-this-in-a-limited-v6-implementation
[^2]: https://github.com/remix-run/react-router/blob/3821d86fc98d1f3762ac48a56510a663876183f8/packages/router/router.ts#L1075

**Comment by jrnxf** [comment]
> React Router supports blocking the navigation by reverting the navigation[1](#user-content-fn-1-61958841f9e36e75e12b3bed09765552)[2](#user-content-fn-2-61958841f9e36e75e12b3bed09765552):
> Are you willing to implement a similar solution? This blocks us and likely other teams from using TanStack Router as we have to requirement to handle dirty state for all navigations.

@tannerlinsley any consideration here - would also love to see this parity if possible

**Comment by rothsandro** [comment]
@jrnxf Blocking back navigation has already been implemented since my comment

**Comment by Jarzka** [comment]
I can also confirm that back navigation works in our app by using `useBlocker`.

**Comment by jrnxf** [comment]
> [@jrnxf](https://github.com/jrnxf) Blocking back navigation has already been implemented since my comment

It will block the first browser back event, but if the user hits the back button more than once, when they opt to "stay" and the reset fn is called, you're not synced properly. For example


User has this navigation stack with a blocker

A -> B -> C -> D -> E

user hits back button 

URL shows: D

now hit back button 2 more times:

URL shows: B

when the reset function is called you'll only be brought back to C in the URL - even though you're visually looking at E
